Fix redundancy in .gw files

GuillaumeBrochu opened the pull request:
See https://github.com/geneweb/geneweb/pull/481

sagotch commented:
This looks nice but I haven't try it yet, sorry.

I would suggest to drop `use_std_fields` support, but I do not understand why are these fields were printed anyway. Any hint?

hgouraud commented:
Should we distinguish `old_gw `from `use_std_fields`?
Both are needed for backwards compatibility with older versions of GeneWeb.

sagotch commented:
IMHO, backward compatibility is important from version `x` to version `x + 1`, but it is okay to break things version `x` to version `x + 2` (we obviously need to be able to use data from the previous installation).

So `old_gw` support should also probably be gone (I do not know how *old* is `old_gw`).

We *could* have a **side** project to translate gw from a version to another, but all that legacy code to ensure backward/forward compatibility leads to messy code, which is more error-prone than the clean one, and which make developers life harder.

Messy and bigger, and probably less efficient code.

GuillaumeBrochu commented:
`old_gw`: to produce a .gw file that is compatible with GW6 (no pevt, no fevt, extra events printed in notes)

`use_std_fields`: NEW proposed feature (this PR) to avoid redundancy (information that is in pevt and fevt is not repeated in standard fields). Only compatible with GW7 (not with GW6), but compatible with GW7 files generated so far with GW7-exp (like the ones generated by Geneanet).

`old_gw `is still required no matter what is decided about this PR (`use_std_fields`)

For further details, please look at https://github.com/geneweb/geneweb/pull/481

hgouraud commented:
The real question is to check that with redundant fields removed, GeneWeb can still read .gw files with those redundant fields!
-> remove redundant fields in gwu, maintain them in gwc

GuillaumeBrochu commented:
> The real question is to check that with redundant fields removed, GeneWeb can still read .gw files with those redundant fields!

It should! Otherwise there is a bug in my PR!

> -> remove redundant fields in gwu, maintain them in gwc

If put my "genealogist hat", I think redundant fields are MUCH more annoying in the .gw file (.gw files are the long term archival container of my genealogy work) than in the internal database (stuff "under the hood" that I don't need to understand).   But your comment is also right : my PR does not remove redundant fields in the internal database. Fabien had the same opinion two years ago : https://github.com/geneweb/geneweb/pull/481#issuecomment-263150420

sagotch commented:
Does this test procedure sounds right? 

1. Export with old version
2. Import with old version
3. md5sum gwb directory
4. Export with new version
5. Import with new version
6. md5sum gwb directory
7. md5sum should be equal

sagotch commented:
Could we drop everything inside `if use_std_fields` block and only keep the `then`?

GuillaumeBrochu commented:
> Does this test procedure sounds right?
> 
> 1. Export with old version
> 2. Import with old version
> 3. md5sum gwb directory
> 4. Export with new version
> 5. Import with new version
> 6. md5sum gwb directory
> 7. md5sum should be equal

I have tested my code with a similar procedure, but I have use `diff` on gw files.

GuillaumeBrochu commented:
> Could we drop everything inside `if use_std_fields` block and only keep the `then`?

No. This code is also required to export in GW6 (`old_gw` switch).

Please note that currently, `use_std_fields` is not the default option, but we could easily change that.

sagotch commented:
I am not likely to be able to test it before the 22nd of october, but I'll try to tests/merge it around this date.

GuillaumeBrochu commented:
@sagotch 

I just have rethink about your comment regarding `use_std_fields` (https://github.com/geneweb/geneweb/pull/685#issuecomment-428227811)

You were right : `use_std_fields` is not strictly required. The `old_gw `switch is enough. 

My initial intent two years ago when I started to work on this was to add an option allowing to generate .gw files like "early releases" of geneweb 7.00-exp does, but this is not strictly required. For illustration purpose, this unnecessary switch correspond to this demonstration code : https://github.com/GuillaumeBrochu/geneweb/commit/b4791007bab38d79195611ca06f1cb52efeb7151

Therefore, I have modified this pull request and remove the `std_fields` switch.

Also, I will make further tests with the grimaldi base and my own base to make sure everything is OK

GuillaumeBrochu commented:
I have a new idea to improve the big code block inside the `if !old_gw` for birth date. Stay tuned.

GuillaumeBrochu committed:
Preliminary rewriting, nothing is really changed

- Add zero_birth_is_required to simplify code
- remove unused code : has_infos_isolated
- has_infos_not_dates, has_infos : order, one parameter per line

GuillaumeBrochu committed:
Fix has_infos_not_dates in gwuLib.ml

Because #image, #apubl, and #apriv is also personal information
as well as #rp, #bs, #ps, #ds, #rs

GuillaumeBrochu commented:
Here we are. This implementation is much better and easier to understand.

Here is my test script **(UPDATED, SEE BELOW)**: 

Up to now, the `diff` at the end of the script is always empty, meaning that no information is lost in the process.

I will also make further tests for the `-old-gw` switch, to make sure it works as expected.

GuillaumeBrochu commented:
I just have:
- fixed a remaining bug with `csrc`.
- updated the test script (see below) to include a test for `-old_gw`
- created a new test base trying to use all fields

Test script:
```
#!/bin/sh
test_dir="/home/user/geneweb/testPR685"
base_dir="$test_dir/bases"
input_base="$test_dir/base_to_test.gw"
cd $base_dir
cp $input_base ./B0.gw
echo ----
echo $input_base
echo --V7--
echo --gwc--
../distribution_master/gw/gwc -nofail -f B0.gw -o B1 > comm.log
echo --gwu--
../distribution_master/gw/gwu B1 -o B1.gw > comm.log
echo --gwc--
../distribution_master/gw/gwc -nofail -f B1.gw -o B2 > comm.log
echo --gwu--
../distribution_master/gw/gwu B2 -o B2.gw > comm.log
../distribution_fix_gwu_redundancy/gw/gwu B2 -o N2.gw > comm.log
echo --gwc--
../distribution_master/gw/gwc -nofail -f N2.gw -o B3 > comm.log
echo --gwu--
../distribution_master/gw/gwu B3 -o B3.gw > comm.log
echo --V6--
echo --gwu--
../distribution_master/gw/gwu B1 -old_gw -o V1.gw > comm.log
echo --gwc--
../distrib-6-08-ocaml-4-xx/gw/gwc -nofail -f V1.gw -o V2 > comm.log
echo --gwu--
../distrib-6-08-ocaml-4-xx/gw/gwu V2 -o V2.gw > comm.log
echo --gwc--
../distribution_master/gw/gwc -nofail -f V2.gw -o B4 > comm.log
echo --gwu--
../distribution_fix_gwu_redundancy/gw/gwu B4 -old_gw -o O4.gw > comm.log
echo --gwc--
../distrib-6-08-ocaml-4-xx/gw/gwc -nofail -f O4.gw -o V4 > comm.log
echo --gwu--
../distrib-6-08-ocaml-4-xx/gw/gwu V4 -o V4.gw > comm.log
echo --diff v7--
diff B2.gw B3.gw | colordiff
echo --diff v6--
diff V2.gw V4.gw | colordiff
echo ----
```

**IMPORTANT NOTE :** this PR (`distribution_fix_gwu_redundancy`) was rebased on #694 to avoid the bug described in #693, while I used #694 for `distribution_master`
 
Test base:
[TC3.gw.txt](https://github.com/geneweb/geneweb/files/2497734/TC3.gw.txt)

**RESULT :** Both diff (V6 and V7) are empty :-)

GuillaumeBrochu committed:
gwu : Only export standard fields in -old_gw mode

To fix redundancy in .gw files

Use dummy text date 0(b) 0(p) 0(d) if required

sagotch commented:
Tested with my base. Seems to work as expected. @GuillaumeBrochu do you need to modify it or can I merge now?

sagotch commented:
Tested with my base. Seems to work as expected. @GuillaumeBrochu do you need to modify it or can I merge now?

sagotch commented:
Tested with my base. Seems to work as expected. @GuillaumeBrochu do you need to modify it or can I merge now?

sagotch commented:
Tested with my base. Seems to work as expected. @GuillaumeBrochu do you need to modify it or can I merge now?

sagotch commented:
@GuillaumeBrochu do you need to update this or can I merge?

GuillaumeBrochu commented:
I do not have pending modifications, it could be the final version, if everyone agree.

sagotch commented:
Sorry about all that posts, github had some troubles this morning...

sagotch commented:
Il semblerait y avoir un problème lors de l'export/import pour ce qui est des indivus sans famille (uniquement référencés comme témoins par exemple).

Apparement un export GW ne conserve pas la date, et du coup à l'import ça donne une date texte : `"(b)"`

dominique95880 commented:
Je crois que "le mal est plus profond" et ne concerne pas que les témoins : 
Je viens de réaliser le test suivant
1° Créer la base de démonstration grimaldi
2° Utiliser gwu pour faire un extrait des ascendants d'Albert et descendants de Rainier, conjoints  inclus et j'obtiens le même résultat pour ce qui concerne les parents de Grace 

Je lance le fichier de commandes suivants
```
Echo on
Cls

Set PgmFil=%ProgramFiles%
Set exedir=%PgmFil%\GeneWeb-test1\gw
Set basedir=%Disque%\mes-documents\Dominique\G‚n‚alogie\GeneWeb bases test1

Set base=%1
If "%base%"=="" Set base=grimaldi

Chdir /D "%basedir%"

"%exedir%\gwu.exe" %base% -o %Temp%\test-grimaldi.gw -a "albert alexandre louis pierre" "0" "grimaldi" -d "rainier louis henri maxence bertrand" "0" "grimaldi" -sp -nn -v > %Temp%\comm.log
"%exedir%\gwc1.exe" -o test-grimaldi -f -stats -cg -v -sep %Temp%\test-grimaldi.gw >> %Temp%\comm.log

Set PgmFil=
Set exedir=
Set basedir=
Set base=
%SystemDrive%
Chdir \
```
Fichier gw obtenu : voir P.J. (à renommer en *.gw)
le problème se situe en ligne 31 définissant les parents de Grace
```
fam KELLY John-Brendan #occu Entrepreneur_de_Travaux_Publics 0(b) 0(d) + MAJER Margaret-Katherine 0(b) 0(d)
fevt
#marr 30/1/1924 #p East_Falls,Philadelphia_County,Pa,UNITED_STATES
end fevt
beg
- f _"Grace"_Patricia #occu Actrice 0(b) 0(d)
end

pevt KELLY _"Grace"_Patricia
#birt 12/11/1929 #p Philadelphia,Philadelphia_County,Pa,UNITED_STATES
#deat 14/9/1982 #p Monaco,MONACO
end pevt
```
[test-grimaldi.txt](https://github.com/geneweb/geneweb/files/3059837/test-grimaldi.txt)

sagotch commented:
Pas encore eu le temps d'analyser ton message mais merci pour ton retour.

Au passage, du coup j'ai proposé un revert du commit qui pose problème ici (https://github.com/geneweb/geneweb/pull/762), je lancerai une discussion autour du format d'archivage `.gw` mais je dois déjà faire un état des lieux avant.

GuillaumeBrochu commented:
En ce qui concerne l'exemple de @dominique95880, le problème ne serait-il pas plutôt que les `pevt` ne sont pas imprimés (comme il le devrait) dans les fichiers .gw lorsqu'on extrait seulement une partie de la base (ascendants et/ou descendants)? 

Ces `pevt` semblent pourtant être imprimés correctement lorsqu'on extrait toute la base (selon les fichiers que j'ai conservés suite à mes tests sur la #685, mais je n'ai pas encore revérifié avec la branche master à jour). Voici ce qui devrait aussi apparaître dans le .gw:

```
pevt KELLY John-Brendan
#birt 4/10/1889 #p Philadelphia,Philadelphia_County,Pa,UNITED_STATES
#deat 20/6/1960 #p Philadelphia,Philadelphia_County,Pa,UNITED_STATES
end pevt
```

```
pevt MAJER Margaret-Katherine
#birt 13/12/1898 #p Philadelphia,Philadelphia_County,Pa,UNITED_STATES
#deat 6/1/1990 #p Philadelphia,Philadelphia_County,Pa,UNITED_STATES
end pevt
```

En outre, est-ce qu'un évènement qui n'est pas dans les champs standards de geneweb 6 (par exemple `#cens`, la liste complète est ici https://geneweb.tuxfamily.org/wiki/gw/fr#.C3.89v.C3.A8nements_personnels) est exporté correctement lorsqu'on extrait seulement une partie de la base (ascendants ou descendants)?

Si jamais vous changez d'idée avec la #762, je pourrais investiguer davantage sur ce problème et tenter de le corriger, si possible, sans faire un revert du commit f7d6b68e3de9c0a7df42690c7edb3c125d833a8a. Entre autres, il faudrait vérifier que la non-impression des `pevt` manquants résulte bien précisément du commit  f7d6b68e3de9c0a7df42690c7edb3c125d833a8a et non d'un problème antérieur ou d'un changement effectué après l'inclusion dudit commit. Ceci pourrait être validé en comparant la #762 (qui a un revert de f7d6b68e3de9c0a7df42690c7edb3c125d833a8a) et la branche master en date du jour (avec le f7d6b68e3de9c0a7df42690c7edb3c125d833a8a inclus), ainsi qu'en comparant un build sur le commit f7d6b68e3de9c0a7df42690c7edb3c125d833a8a avec un build sur le commit qui précède f7d6b68e3de9c0a7df42690c7edb3c125d833a8a.

GuillaumeBrochu commented:
P.S.

Voici l'historique pour le fichier gwuLib.ml (divisé en 5 parties puisque git ne réussit pas à considérer les déplacements de fichiers dans sa génération de l'historique):

1: https://github.com/geneweb/geneweb/commits/master/lib/gwuLib.ml

2: https://github.com/geneweb/geneweb/commits/86b9150a5420f62500fbd36fb723637a62926f08/internal/gwuLib.ml

3: https://github.com/geneweb/geneweb/commits/8ed7e577b5d8244965be233514684371d2d1572c/lib/gwuLib.ml

4: https://github.com/geneweb/geneweb/commits/7c4e1767a80ca1f6f2f341aa61a60f2340aa98ec/src/gwuLib.ml

5: https://github.com/geneweb/geneweb/commits/dbb75a7ca091baa22ba86c86538f1f5593f29b12/src/gwu.ml

GuillaumeBrochu commented:
La fonction `print_pevents_for_person`
https://github.com/geneweb/geneweb/blob/d54ad1d1412ecc5e799217d7d9ee4f8ea46c0cd6/lib/gwuLib.ml#L708-L720

est appelée deux fois:
ici: https://github.com/geneweb/geneweb/blob/d54ad1d1412ecc5e799217d7d9ee4f8ea46c0cd6/lib/gwuLib.ml#L731-L743
et ici: https://github.com/geneweb/geneweb/blob/d54ad1d1412ecc5e799217d7d9ee4f8ea46c0cd6/lib/gwuLib.ml#L1333-L1355

GuillaumeBrochu commented:
Je vois potentiellement deux problèmes potentiels ici (à confirmer):
https://github.com/geneweb/geneweb/blob/d54ad1d1412ecc5e799217d7d9ee4f8ea46c0cd6/lib/gwuLib.ml#L1733-L1776

1. Il faudrait utiliser `has_infos` ou `has_infos_not_in_events` (à déterminer) au lieu de tout expliciter

2. Imprimer aussi les `pevt` ? (à confirmer)

P.S. : Voici le commit initial: https://github.com/geneweb/geneweb/commit/c41bc27ed26bc854875e4e3c3096a9ce6cfda040

GuillaumeBrochu commented:
Je crois qu'il manque aussi l'impression des `pevt` ici (à confirmer):
https://github.com/geneweb/geneweb/blob/d54ad1d1412ecc5e799217d7d9ee4f8ea46c0cd6/lib/gwuLib.ml#L1357-L1376

GuillaumeBrochu commented:
Une dernière observation pour ce soir:

Les `pevt` sont souvent imprimés juste après les notes de la personne. Il suffit donc de chercher `print_notes` et `print_notes_for_person` afin de déduire où devraient aussi être imprimés les `pevt` avec `print_pevents` ou `print_pevents_for_person`

sagotch commented:
Pour ce qui est du revert de #762 , il n'y a pas de problème pour revert le revert dans une autre PR. En attendant, je crois qu'il est plus sage de revenir à une version qui perd moins d'infos.

Effectivement, comme je le disais dans la PR:
> Note: gwu still may loose some info since this same witness is no exported except on a wit line (i.e. its pevents are not exported, so secondary events are lost). If this is really what happens, it is probably the case since a long time, and maybe it is how it has always been.

Donc effectivement, il y a à priori un problème de pevent perdu.

Je n'ai pas eu le temps d'investiguer en profondeur, mais d'après ce que j'ai vu

```ocaml
    let print_witness oc base gen p =
      Printf.fprintf oc "%s %s%s" (correct_string base (get_surname p))
        (correct_string base (get_first_name p))
        (if get_new_occ p = 0 then ""
         else "." ^ string_of_int (get_new_occ p));
      if Array.length (get_family p) = 0 && get_parents p = None &&
         not gen.mark.(Adef.int_of_iper (get_key_index p))
      then
        begin
          gen.mark.(Adef.int_of_iper (get_key_index p)) <- true;
          if has_infos base p then print_infos oc base false "" "" p
          else Printf.fprintf oc " 0";
          begin match sou base (get_notes p) with
            "" ->
              if put_events_in_notes base p then
                gen.notes_pl_p <- p :: gen.notes_pl_p
          | _ -> gen.notes_pl_p <- p :: gen.notes_pl_p
          end;
          if get_pevents p <> [] then
            gen.pevents_pl_p <- p :: gen.pevents_pl_p
        end
```

Si je ne dis pas de bêtise, le `gen.pevents_pl_p <- p :: gen.pevents_pl_p` ne sert à rien ici car `gen.pevents_pl_p` n'est plus utilisé par la suite. J'ai l'impression que l'endroit qui boucle sur `gen.pevents_pl_p` devrait se faire justement dans une loop, et pas un `List.fold_right` + `List.iter`.

```ocaml
    let print_pevents oc base gen ml =
      let pl =
        List.fold_right get_persons_with_pevents ml gen.pevents_pl_p
      in
      let pl =
        List.fold_right
          (fun p pl -> if list_memf eq_key p pl then pl else p :: pl) pl []
      in
      List.iter
        (fun p ->
           if gen.per_sel (get_key_index p) then
             print_pevents_for_person oc base gen p)
        pl
```

hgouraud commented:
En lisant gwuLib, je note en passant un problème qu'on a eu avec Roglo 
La fonction read_file_contents ouvre les fichiers notes des wizards, mais ne les ferme pas!
Roglo perdait les dernières notes de wizard à chaque gwu! (plus de 200 wizards)
Christian à traité le pb en augmentant la limite du nombre de fichiers ouverts, et de mon côté, j'ai rajouté un close ic en fin de fonction (sur la 6.08 spécial Roglo!!)(bretelles et ceinture!!)

GuillaumeBrochu commented:
@sagotch 

> ici car `gen.pevents_pl_p` n'est plus utilisé par la suite.

Si j'ai bien compris, c'est ici que le contenu de `gen.pevents_pl_p` est utilisé : https://github.com/geneweb/geneweb/blob/d54ad1d1412ecc5e799217d7d9ee4f8ea46c0cd6/lib/gwuLib.ml#L731-L734

À propos de gen, je vois aussi ceci (est-ce que cela est cohérent avec le reste du code de gwuLib?): 

https://github.com/geneweb/geneweb/blob/d54ad1d1412ecc5e799217d7d9ee4f8ea46c0cd6/lib/gwuLib.ml#L425-L434

https://github.com/geneweb/geneweb/blob/d54ad1d1412ecc5e799217d7d9ee4f8ea46c0cd6/lib/gwuLib.ml#L1675-L1687

Il reste à bien comprendre ce code et à déterminer si le problème est relié à la façon dont `gen.pevents_pl_p` est utilisé dans `print_witness` ET/OU qu'il manque des `print_pevents` ou des `print_pevents_for_person` à quelques endroits dans le code (ce qui me semble de plus en plus clair). Il y a peut-être deux problèmes en fait ...

sagotch commented:
J'ai surtout l'impression que tout est un peu fait pele-mele.

Une façon qui me semble plus simple de procéder, bien que à priori plus longue (mais pour moi c'est ok car gwu n'a pas à être utilisé souvent et donc peut souffrir d'un manque d'optimisation si c'est au profit de la fiabilité et de la simplicité).

Faire une première passe de selection des individu et familles à exporter. Ensuite, exporter tout ce petit monde avec toutes les infos qui les concernent.

Là j'ai l'impression que l'impression et la selection s'opèrent plus ou moins en meme temps et se combinent mal.
